core/arch/instr_shared.c

```diff
@@ -503,16 +536,26 @@ instr_set_num_opnds(dcontext_t *dcontext, instr_t *instr, int instr_num_dsts,
         CLIENT_ASSERT_TRUNCATE(instr->num_dsts, byte, instr_num_dsts,
                                "instr_set_num_opnds: too many dsts");
         instr->num_dsts = (byte)instr_num_dsts;
-        instr->dsts = (opnd_t *)heap_alloc(
-            dcontext, instr_num_dsts * sizeof(opnd_t) HEAPACCT(ACCT_IR));
+        if (TEST(INSTR_IS_NOALLOC_STRUCT, instr->flags)) {
+            instr_noalloc_t *noalloc = (instr_noalloc_t *)instr;
```

Comment:
If find this cast and others messy tbh. I thought instr_noalloc_t was enclosing instr_t and cannot be used vice-versa as well. This makes the code hard to read overall. Of course, this is just my personal opinion, and may be wrong here.

What I am thinking is to create separate API entries for stack/heap allocated instr and have these functions internally make use of the same code. I know this can be a bit of a pain but given the fact that we are avoiding unions, I think this approach can help mitigate type confusion and improve clarity.

Reply:
This is essentially a sub-class: extending the instr_t struct with additional fields and passing a pointer to the base class around.

Reply:
Yes, I understood what you did here. Perhaps, if worded differently, why didn't you modify decode() to also check the instr flag for instr_noalloc_t? Requiring the user to fetch the base in your test case implies non-interchangeability which contrasts with the code above.

Reply:
With no compiler support for overloads, we can't modify decode() to solve this: the compiler complains about the type passed in.  We'd have to make a separate function decode_noalloc().  We'd have to make a hundred separate functions.  IMHO it is a nicer interface to instead pass in instr_from_noalloc(&noalloc) (proposed name for your suggested helper to return &noalloc.instr) to the existing functions, with a little hackiness hidden in the implementation for the poor man's subclass.

Reply:
Thanks, I understand the the overkill and the fact that we r trying to minimise changes. I got the PR to run on my machine. Actually, what I meant, is the passing of &noalloc to instr_get_dst(), which according to the docs accepts instr_t but the check is circumvented due to it being a macro. Perhaps modify the docs for those functions?

Reply:
Interesting, I just used it without thinking and the compiler did not complain.  You mean this:
```C
    ASSERT(opnd_get_reg(instr_get_dst(&noalloc, 0)) == DR_REG_X0);
```

I will change it `s/&noalloc/instr_from_noalloc(&noalloc)/`.  I guess that means the suggested use is:

```C
instr_noalloc_t noalloc;
instr_noalloc_init(dcontext, &noalloc);
instr_t *instr = instr_from_noalloc(&noalloc);
<now use instr everywhere; no need to free at the end, but still do a reset to re-init if doing repeated decode into it or sthg>
```

WDYT?

Reply:
Yea, much better - essentially avoid interchanging between instr and instr_noalloc at the API level or any other types for that matter because that's when type confusion bugs start to crop up.

Reply:
Implemented.